### DISCUSSION TITLE
Can multiple `autoload.php`'s be required at runtime by design ?

### DISCUSSION THREAD

**Discussion by MatteoOreficeGS** [composer/composer#12087] [General]
Hello everyones, I would ask you :

> is composer designed to allow multiple require of autoload.php in the same runtime / execution flow? I mean multiple `autoload.php` generated by indipendent `composer.json`

What kind of issues could arise ?

- build time conflict ? because composer update / require is run in two separate times are random names generated in vendor/composer/autoload* guaranteed with respect to name collisions?
- indirect dependencies conflict ? the two composer.json have different roots but some indirect deps could be shared between them

I discovered a particular layout in a **project** where two different `autoload.php` are included during a regular API call from two indipendent vendor locations, and I am asking myself if this  undocumented feature is 100% safe

I cannot find in the doc a note about 

IMHO I would use a different approach with repo `path` but maybe other people know something more about composer internal to consider this approach safe

Thanks !!!

**Comment by stof** [comment]
The autoloader files themselves are safe if the 2 generated autoloaders are using a different [`autoloader-suffix`](https://getcomposer.org/doc/06-config.md#autoloader-suffix) (that's the exact reason why this suffix exists).

However, if you load 2 sets of packages in the same PHP process and they have some packages in common but at different versions, this will cause issues because PHP cannot load 2 different definitions for the same class name in a process (note that this is totally unrelated to Composer. It is how PHP works).
If you are lucky, one of the installed package will consistently win (and then, it is a question whether this version is compatible with everything). If you are unlucky, you might end up having some classes loaded from one version of the package and others loaded from the other version (for instance when some classes exist in only one of those versions).

**Reply by MatteoOreficeGS** [reply]
Hello @stof,  I didn't know about the `autoload-suffix` , I think we need to integrate in our config ASAP

I supposed also what you are saying about PHP class loading behaviour, and how it is unrelated to composer, but I was in doubt composer added any constraints.

thanks for the clarification !
